src/parser/program.rs

```diff
@@ -0,0 +1,157 @@
+use libc::sys::types::Pid;
+use serde::Deserialize;
+use serde_yaml::Error;
+use std::{collections::HashMap, fmt::Display, fs::File};
+
+use super::ParseError;
+
+#[derive(Debug, Deserialize)]
+pub struct ParsedConfig {
+    pub programs: HashMap<String, ParsedProgram>,
+}
+
+pub struct Config {
+    pub programs: Vec<Program>,
+}
+
+#[derive(Debug, Deserialize)]
+#[allow(dead_code)] // TODO: remove this
+pub struct EnvVar {
+    key: String,
+    value: String,
+}
+
+#[derive(Debug, Deserialize)]
+pub struct ParsedProgram {
+    cmd: String,
+    numprocs: Option<u32>,
+    workingdir: Option<String>,
+    autostart: Option<bool>,
+    exitcodes: Option<Vec<u8>>, // check for valid codes (%256)
+    startretries: Option<u32>,
+    starttime: Option<u32>,
+    stopsignal: Option<String>, // check for valid signal
+    stoptime: Option<u32>,
+    stdout: Option<String>,
+    stderr: Option<String>,
+    env: Option<HashMap<String, String>>,
+}
+
+#[derive(Debug)]
+#[allow(dead_code)] // TODO: remove this
+pub struct Program {
+    name: String,
+    pids: Vec<Pid>,
+    cmd: String,
+    numprocs: u32,
+    workingdir: String,
+    autostart: bool,
+    exitcodes: Vec<u8>, // check for valid codes (%256)
+    startretries: u32,
+    starttime: u32,
+    stopsignal: String, // check for valid signal
+    stoptime: u32,
+    stdout: String,
+    stderr: String,
+    env: Vec<EnvVar>,
+}
+
+impl ParsedConfig {
+    pub fn new(file: File) -> Result<Self, Error> {
+        let new_config = serde_yaml::from_reader(file)?;
+        Ok(new_config)
+    }
+}
+
+impl From<ParsedConfig> for Config {
+    fn from(origin: ParsedConfig) -> Self {
+        Self {
+            programs: origin
+                .programs
+                .into_iter()
+                .fold(vec![], |mut acc, (key, value)| {
+                    acc.push(Program::from(key, value));
+                    acc
+                }),
+        }
+    }
+}
+
+impl Program {
+    fn from(name: String, origin: ParsedProgram) -> Self {
```

Comment:
Make this a try_from (Implement the trait TryFrom) and run the check of stopsignal here so that it is impossible to create an instance Program with invalid data